File: ROS2-basic/chapter5/chapter5_ws/src/demo_python_tf/demo_python_tf/dynamic_tf_broadcaster.py
import math
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster             # 动态坐标发布器
from geometry_msgs.msg import TransformStamped       # 消息接口
from tf_transformations import quaternion_from_euler # 欧拉角转四元数函数

class DynamicTFBroadcaster(Node):
    def __init__(self):
        super().__init__('dynamic_tf_broadcaster')
        self.tf_broadcaster_ = TransformBroadcaster(self)
        self.timer_ = self.create_timer(0.01, self.publish_dynamic_tf)

    def publish_dynamic_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "camera_link"
        transform.child_frame_id = "bottle_link"

        # 设置平移部分
        transform.transform.translation.x = 0.2
        transform.transform.translation.y = 0.0
        transform.transform.translation.z = 0.5
 
        # 欧拉角转四元数 q=x,y,z,w
        q = quaternion_from_euler(math.radians(180), 0, 0)
 
        # 旋转部分进行赋值
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]
 
        # 动态坐标关系发布
        self.tf_broadcaster_.sendTransform(transform)
        self.get_logger().info(f'发布 TF: {transform}')

def main():
    rclpy.init()
    node = DynamicTFBroadcaster()
    rclpy.spin(node)
    rclpy.shutdown()

File: ROS2-basic/chapter5/chapter5_ws/src/demo_cpp_tf/src/dynamic_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/transform_broadcaster.h"
#include <memory>
#include <chrono>


class DynamicTFBroadcaster : public rclcpp::Node
{
	public:
		DynamicTFBroadcaster() : Node("dynamic_tf_broadcaster")
		{
			broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
			timer_ = this->create_wall_timer(std::chrono::microseconds(10), std::bind(&DynamicTFBroadcaster::publish_tf, this));
		}
		void publish_tf()
		{
			geometry_msgs::msg::TransformStamped transform;
			transform.header.stamp = this->get_clock()->now();
			transform.header.frame_id = "map";
			transform.child_frame_id = "base_link";
			transform.transform.translation.x = 2.0;
			transform.transform.translation.y = 3.0;
			transform.transform.translation.z = 0.0;
			tf2::Quaternion quat;
			quat.setRPY(0, 0, 30 * M_PI / 180);              // 弧度制欧拉角转四元数
			transform.transform.rotation = tf2::toMsg(quat); // 四元数转ROS消息类型
			broadcaster_->sendTransform(transform);
		}

	private:
		std::shared_ptr<tf2_ros::TransformBroadcaster> broadcaster_;
		rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<DynamicTFBroadcaster>());
	rclcpp::shutdown();
	return 0;
}

